builtin/mainmenu/serverlistmgr.lua

```diff
@@ -0,0 +1,234 @@
+--This program is distributed in the hope that it will be useful,
+--but WITHOUT ANY WARRANTY; without even the implied warranty of
+--MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+--GNU Lesser General Public License for more details.
+--
+--You should have received a copy of the GNU Lesser General Public License along
+--with this program; if not, write to the Free Software Foundation, Inc.,
+--51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+serverlistmgr = {}
+
+--------------------------------------------------------------------------------
+local function order_server_list(list)
+	local res = {}
+	--orders the favorite list after support
+	for i = 1, #list do
+		local fav = list[i]
+		if is_server_protocol_compat(fav.proto_min, fav.proto_max) then
+			res[#res + 1] = fav
+		end
+	end
+	for i = 1, #list do
+		local fav = list[i]
+		if not is_server_protocol_compat(fav.proto_min, fav.proto_max) then
+			res[#res + 1] = fav
+		end
+	end
+	return res
+end
+
+local public_downloading = false
+
+--------------------------------------------------------------------------------
+function serverlistmgr.sync()
+	if not serverlistmgr.servers then
+		serverlistmgr.servers = {{
+			name = fgettext("Loading..."),
+			description = fgettext_ne("Try reenabling public serverlist and check your internet connection.")
+		}}
+	end
+
+	if public_downloading then
+		return
+	end
+	public_downloading = true
+
+	core.handle_async(
+		function(param)
+			local http = core.get_http_api()
+			local url = ("%s/list?proto_version_min=%d&proto_version_max=%d"):format(
+				core.settings:get("serverlist_url"),
+				core.get_min_supp_proto(),
+				core.get_max_supp_proto())
+
+			local response = http.fetch_sync({ url = url })
+			if not response.succeeded then
+				return {}
+			end
+
+			local retval = core.parse_json(response.data)
+			return retval and retval.list or {}
+		end,
+		nil,
+		function(result)
+			public_downloading = nil
+			local favs = order_server_list(result)
+			if favs[1] then
+				serverlistmgr.servers = favs
+			end
+			core.event_handler("Refresh")
+		end
+	)
+end
+
+--------------------------------------------------------------------------------
+local function get_favorites_path()
+	local base = core.get_user_path() .. DIR_DELIM .. "client" .. DIR_DELIM .. "serverlist" .. DIR_DELIM
+	return base .. core.settings:get("serverlist_file")
+end
+
+--------------------------------------------------------------------------------
+local function save_favorites(favorites)
+	local filename = core.settings:get("serverlist_file")
+	if filename:sub(#filename - 3):lower() == ".txt" then
+		core.settings:set("serverlist_file", filename:sub(1, #filename - 4) .. ".json")
+	end
+
+	local path = get_favorites_path()
+	core.create_dir(path)
+	core.safe_file_write(path, core.write_json(favorites))
+end
+
+--------------------------------------------------------------------------------
+function serverlistmgr.read_legacy_favorites(path)
```

Comment:
Should this delete the legacy file? I guess it's cleaner to do so, and favourites aren't exactly vital to keep a backup of

Reply:
Sounds fine to me